### ISSUE TITLE
DNS Server Replacement Thoughts

### ISSUE THREAD

**Issue by bexelbie** [vagrant-landrush/landrush#167]
During the committer meeting we talked briefly about our challenges with DNS services under Windows.  Currently our reliance on a Ruby DNS server puts us at risk for a lot of issues with dependencies.  Currently, the libraries that the DNS server rely on do not properly fork processes under Windows.

One idea that came out of this was to move to a more self-contained DNS server that would be more immune to dependency and OS issues.  In the meeting, the idea of finding a Go-lang DNS server was talked about, however a quick search didn't find one.

I've tried to make a more comprehensive search and found the following options (in no particular order):

* bind: https://www.isc.org/downloads/bind/
* posadis: http://posadis.sourceforge.net/
* MaraDNS: http://maradns.samiam.org/index.html
* lbnamed: http://web.stanford.edu/~riepel/lbnamed/
* yadifa: http://www.yadifa.eu/home
* skydns: https://github.com/skynetservices/skydns

My Commentary (YMMV):
* bind: The mackdaddy of DNS servers.  It is probably to heavy for our needs.  It also doesn't enjoy a gr
eat reputation for configuration and startup, especially under Windows.  However, it may be worth giving
real though too.
* posadis: This daemon seems to be recommended but also seems to be dead.
* MaraDNS: This daemon is in maintenance mode.  It doesn't seem to have a community of contributors, which concerns me.  However it seems to fit the bill.
* lbnamed: This daemon is written in perl so it is just begging to have the same problems our ruby soluti
on has.  But ... perl!!!!
* yadifa: A seemingly solid daemon that, like bind, may be too heavy.
* skydns: The sexy Go-Lang champion.  All the cool cloud kidz love it.  However, it requires etcd which m
eans we have to spawn two daemons, not one.  The side benefit is that it probably has an outsized community to draw on ...

I have not done any serious exploration of how hard this would be to add.  @hferentschik has dome some serious lifting in PRs and we could probably just make Windows support beta and move on.  But we will still be at the mercy of ruby for a while.

Any opinions/thoughts? @phinze @strzibny @fh @ahpook @njam

**Comment by hferentschik** [comment]
> we could probably just make Windows support beta and move on.

+1 I see no problem with the Ruby DNS server. Yes we need to downgrade Ruby DNS, but that is a minor issue for me. The version we are using has been used for a long time and is known to be stable.

>  But we will still be at the mercy of ruby for a while.

You make it sounds like this is a bad thing? Nothing wrong with Ruby imo. 

> One idea that came out of this was to move to a more self-contained DNS 

-1 from me. The beauty is that we ca drive everything from Ruby and stay in a single framework/programming language. If you rely on a different type of DNS server you need yet another set of binaries (for each OS), but additional install routines, etc. To me that makes no sense at all to rely on a independent DNS server which needs to be installed separately. Ruby has all the means to write proper portable code which runs on all OSes without problems. The issue is that a lot of libraries are not intended to be cross OS or if they are they are often not tested properly for all OSes. 

Regarding Ruby DNS. The problem with the latest version is not the use of `fork`, but rather the use of the celluoid framework to implement the request handling part. The code actually runs on Windows. The problem is just that it deadlocks. Potentially this could be fixed. 

Last but not least, if nothing else, we can easily write our own simple DNS server. If you look at the Ruby DNS code, it is just a few lines of code around some other framework (Celluloid or event-machine) to handle incoming request. This won't be a production grade DNS server of course, but I don't see Landrush in this space anyway.

Bottom line, imo Landush and its DNS server it starts/manages should stay Ruby.

**Comment by strzibny** [comment]
Completerly agree with @hferentschik. I would rather maintain old version/fork for Ruby DNS.

**Comment by bexelbie** [comment]
@hferentschik that works for me.  I wanted to keep that conversation rolling through the committers and move us forward.  Since three of us are on board with the move to Windows == Beta .. then I think we can do it.  Now let's merge some patches to that effect :)

**Comment by hferentschik** [comment]
> that works for me.

cool

>  Since three of us are on board with the move to Windows == Beta .. then I think we can do it. 

+1

**Comment by hferentschik** [comment]
Closing this for now. Switching to a non Ruby DNS solution is not an option for now.
